BUILD.gn

```diff
@@ -0,0 +1,204 @@
+#  Copyright (c) 2019, The OpenThread Authors.
+#  All rights reserved.
+#
+#  Redistribution and use in source and binary forms, with or without
+#  modification, are permitted provided that the following conditions are met:
+#  1. Redistributions of source code must retain the above copyright
+#     notice, this list of conditions and the following disclaimer.
+#  2. Redistributions in binary form must reproduce the above copyright
+#     notice, this list of conditions and the following disclaimer in the
+#     documentation and/or other materials provided with the distribution.
+#  3. Neither the name of the copyright holder nor the
+#     names of its contributors may be used to endorse or promote products
+#     derived from this software without specific prior written permission.
+#
+#  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+#  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+#  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+#  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+#  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+#  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+#  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+#  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+#  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+#  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+#  POSSIBILITY OF SUCH DAMAGE.
+#
+
+static_library("lib-ot-core") {
+
+ cflags_cc = [ "-Wno-non-virtual-dtor" ]
+
+ sources = [
```

Comment:
A question/suggestion: Is there any solution to have this list be auto generated (more generic), e.g., include all `cpp` files in `src/core/api`, and `src/core/common` and ...)?
If we can do this,  I think it would be more generic and robust against file renames and new src files added later.

Thoughts?

Reply:
Currently we are matching how things are done in Android.mk file. 

It is possible to provide wildcards to include all cpp files within a given directory.  But not sure whether we want to do that, as it may produce unintentional side effects for Fuchsia specific directory.  For example, we have explicitly removed posix specific files which were there in Android.mk.  

This is just a starting point, and the file is expected to evolve anyway when new files/stubs are added for Fuchsia. We can make the change now (to include all cpp files within a directory now), or defer it to wait for things to progress a little bit.

Reply:
Currently we are matching how things are done in Android.mk file.

It is possible to provide wildcards to include all cpp files within a given directory. But not sure whether we want to do that, as it may produce unintentional side effects for Fuchsia specific directory. For example, we have explicitly removed posix specific files which were there in Android.mk.

This is just a starting point, and the file is expected to evolve anyway when new files/stubs are added. We can make the change now (to include all cpp files within a directory now), or defer it to wait for things to progress a little bit.

Reply:
Currently we are matching how things are done in Android.mk file.
Also, we are excluding some files within src/core/api , like : border_agent_api.cpp, similar to Android.mk

It is possible to provide wildcards to include all cpp files within a given directory. But not sure whether we want to do that right now.

This is just a starting point, and the file is expected to evolve anyway when new files/stubs are added. We're just matching Android.mk as a starting point and make the choice consciously of what files need to be included as required later.

Reply:
I would still suggest that if we can use wildcards for certain folders, it would be great to include it from beginning and test it.

We can use the wildcards for certain core folders (the ones we know we always want to include all their src files). And for some other folders use a specific list. 

In general, regarding the OT core files/folder, we would want to include every file in a folder (there are `#if` checks at the top of files if certain features are not enabled), e.g., I see no advantage to not include the `border_agent_api.cpp`.

The main reason for this is to make it easier when new files are added/removed in code or there is a file rename..We want to reduce the number of makefiles/build files that developers need to maintain/update (makes it easier on devs and easier to test changes on travis). This also helps avoid the situation where unused/disabled features would go unnoticed if they are not included in the build/makefiles (we have faced this with andriod.mk in past where enabling feature later, caused build errors that certain src files were missing since they were not updated in andriod.mk during a change).

Reply:
It should be OK to always include all source files in `src/core`. Everything should be appropriately feature flagged. We intentionally support this to simplify maintenance of other build systems.

Reply:
Sorry, I have a correction.   There doesn't seem to be a way to provide source names to include with wild cards. 

When I looked up help earlier (https://gn.googlesource.com/gn/+/master/docs/reference.md#how-to-use-patterns) , I thought patterns are used to include the source files.  However the example in the help shows how to filter out files matching a pattern (ie just provides an exclude list). I have also cross checked that Fuchsia tree which heavily uses GN lists all the source files in the BUILD.gn hierarchy. 

So we're probably stuck with listing out all the files explicitly.

Reply:
I'm still checking to see if there is some trick to make it work, but surely there is no straight forward way.

Reply:
After searching more, I don't think there is a way to provide wildcards for adding files.
I have added all source files under src/core to the BUILD.gn file as suggested above.

Reply:
Based on above, I'm marking the conversation as resolved, please reopen if there is anything additional required.